zio/zngio/scanner.go

```diff
@@ -16,11 +18,83 @@ import (
 
 // zngScanner implements scanner.Scanner.
 type zngScanner struct {
-	ctx    context.Context
-	reader *Reader
-	filter filter.Filter
-	span   nano.Span
-	stats  scanner.ScannerStats
+	ctx          context.Context
+	reader       *Reader
+	filter       filter.Filter
+	span         nano.Span
+	stats        scanner.ScannerStats
+	stringFinder *stringFinder
+}
+
+var _ scanner.ScannerAble = (*Reader)(nil)
+
+func (r *Reader) NewScanner(ctx context.Context, filterExpr ast.BooleanExpr, s nano.Span) (scanner.Scanner, error) {
+	var f filter.Filter
+	var sf *stringFinder
+	if filterExpr != nil {
+		var err error
+		if f, err = filter.Compile(filterExpr); err != nil {
+			return nil, err
+		}
+		p, err := requiredPattern(filterExpr)
+		if err != nil {
+			return nil, err
+		}
+		if len(p) > 1 {
+			// Profitable if pattern is longer than one byte.
+			sf = makeStringFinder(p)
+		}
+	}
+	return &zngScanner{ctx: ctx, reader: r, filter: f, span: s, stringFinder: sf}, nil
+}
+
+// requiredPattern searches e for a required pattern of maximum length. A
+// required pattern for e is a byte sequence that must be present in any buffer
+// that contains the ZNG encoding of a record matching e.
+func requiredPattern(e ast.BooleanExpr) (string, error) {
+	switch e := e.(type) {
+	case *ast.LogicalAnd:
+		left, err := requiredPattern(e.Left)
+		if err != nil {
+			return "", err
+		}
+		right, err := requiredPattern(e.Right)
+		if err != nil {
+			return "", err
+		}
+		if len(right) > len(left) {
+			return right, nil
+		}
+		return left, nil
+	case *ast.CompareAny:
+		return requiredPatternForCompare(e.Comparator, e.Value)
+	case *ast.CompareField:
+		return requiredPatternForCompare(e.Comparator, e.Value)
+	default:
+		return "", nil
+	}
```

Comment:
ast.Search semantics make it tough to handle. We probably need to allow for two patterns, only one of which must match. And how to handle the case folding in filter.searchString is unclear.

Reply:
This seems worthy of being added as comment in the .go file.

Reply:
Done. #1128 tracks the work.